src/core/thread/mle.cpp

```diff
@@ -4418,6 +4428,66 @@ void Mle::DelayedResponseMetadata::RemoveFrom(Message &aMessage) const
     OT_UNUSED_VARIABLE(error);
 }
 
+#if OPENTHREAD_CONFIG_TMF_DUA_REGISTRATION_ENABLE
+void Mle::UpdateDomainUnicastAddress(BackboneRouter::Leader::DomainPrefixState aState)
+{
+    const otIp6Prefix *prefix;
+    bool               attempt = false;
+
+    if (aState == BackboneRouter::Leader::kDomainPrefixRemoved ||
+        (aState == BackboneRouter::Leader::kDomainPrefixRefreshed))
+    {
+        Get<ThreadNetif>().RemoveUnicastAddress(mDomainUnicastAddress);
+    }
+
+    prefix = Get<BackboneRouter::Leader>().GetDomainPrefix();
+
+    VerifyOrExit(prefix != NULL, OT_NOOP);
+
+    if (aState == BackboneRouter::Leader::kDomainPrefixRefreshed)
+    {
+        if (!mDomainUnicastAddress.mValid)
+        {
+            // Will try to generate IID again as the prefix changes.
+            attempt = true;
+        }
+        else
+        {
+            // Cache IID generated to reuse.
+            Ip6::Address address = mDomainUnicastAddress.GetAddress();
+
+            mDomainUnicastAddress.GetAddress().Clear();
+            mDomainUnicastAddress.GetAddress().SetPrefix(prefix->mPrefix.mFields.m8, prefix->mLength);
+            mDomainUnicastAddress.GetAddress().SetIid(address.GetIid());
+            mDomainUnicastAddress.mPrefixLength = prefix->mLength;
+
+            Get<ThreadNetif>().AddUnicastAddress(mDomainUnicastAddress);
+        }
+    }
+
+    if (aState == BackboneRouter::Leader::kDomainPrefixAdded || attempt)
+    {
+        mDomainUnicastAddress.GetAddress().Clear();
+        mDomainUnicastAddress.GetAddress().SetPrefix(prefix->mPrefix.mFields.m8, prefix->mLength);
+
+        if (Get<Utils::Slaac>().GenerateIid(mDomainUnicastAddress, NULL, 0, &mDadCounter,
+                                            false /* not allow random IID if no valid IID */) == OT_ERROR_NONE)
```

Comment:
any background on why not allow random IID ?

Reply:
now mle component owns mDadCounter, which is used for Duplicate detection, I think mle should control and ensure the relationship of IID and mDadCounter.

Reply:
`SLAAC` only generates random IID after failed using `mDadCounter` counter. In such extreme situation, would it better to just use a random IID than to not having DUA?

Reply:
1.2 Specification 5.23.1 says

> Domain Unicast Address (DUA), an IPv6 address for a Thread
> Device that remains stable even after roaming to a new Thread Network configured with the same Thread Domain information

I think the determinacy is expected more than a random one which is unpredictable. Besides, note that existing SLAAC utils tries up to 1024 to avoid the Reserved IID which is just a limited sets

https://github.com/openthread/openthread/blob/master/src/core/net/ip6_address.cpp#L158-L161

I would take a random one to be a `nice to have` feature in some extreme situations that all the IID generated by all the `1024` generates are reserved ones, and the random IID could stored to ensure the consistency when roaming.

This PR might not be so perfection to handle all the possible scenarios in some view, however could we focus on the majority scenarios when review and leave enhancements on some possible extreme scenarios in the future?

Reply:
Fine to not use random IID since the situation hardly ever happen.